common/transport/amqp/src/amqp.ts

```diff
@@ -135,6 +88,280 @@ export class Amqp {
+            let senderFsm = new SenderLink(endpoint, linkOptions, this._amqp);
+            this._senders[endpoint] = senderFsm;
+            const attachErrorHandler = (err) => {
+              debug('sender link error event while attaching: ' + endpoint + ': ' + err.toString());
+              delete(this._senders[endpoint]);
+              done(err);
+            };
+
+            this._senders[endpoint].on('error', attachErrorHandler);
+            debug('attaching sender link for: ' + endpoint);
+            this._senders[endpoint].attach((err) => {
+              if (err) {
+                debug('failed to attach sender link: ' + endpoint + ': ' + err.toString());
+                done(err);
+              } else {
+                this._senders[endpoint].removeListener('error', attachErrorHandler);
+                this._senders[endpoint].on('error', (err) => {
+                  debug('error on sender link: ' + endpoint + ': ' + err.toString());
+                  delete(this._senders[endpoint]);
+                });
+                debug('sender link attached: ' + endpoint);
+                done(null, this._senders[endpoint]);
+              }
+            });
+          },
+          detachReceiverLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback<any>): void => {
+            if (!this._receivers[endpoint]) {
+              this._safeCallback(detachCallback);
+            } else {
+              this._detachLink(this._receivers[endpoint], (err?) => {
+                delete(this._receivers[endpoint]);
+                this._safeCallback(detachCallback, err);
+              });
+            }
+          },
+          detachSenderLink: (endpoint: string, detachCallback: GenericAmqpBaseCallback<any>): void => {
+            this._detachLink(this._senders[endpoint], (err?) => {
+              delete(this._senders[endpoint]);
+              this._safeCallback(detachCallback, err);
+            });
+          }
+        },
+        disconnecting: {
+          _onEnter: (disconnectCallback, err) => {
+            const disconnectFunc = err ? (callback) => {
+              callback();
+            } : (callback) => {
+              this._amqp.removeAllListeners('disconnected');
+              /*Codes_SRS_NODE_COMMON_AMQP_16_004: [The disconnect method shall call the done callback when the application/service has been successfully disconnected from the service] */
+              this._amqp.disconnect().then(() => {
+                callback();
+                return null;
+              }).catch((err) => {
+                callback(err);
+              });
+            };
+
+            const detachFunc = err ? (link, callback) => {
+              if (link) {
+                link.forceDetach(err);
+              }
+              callback();
+            } : (link, callback) => {
+              if (link) {
+                link.detach(callback);
+              } else {
+                callback();
+              }
```

Comment:
If we're disconnecting because of an error, don't we still want to detach all links?

Reply:
Also, shouldn't we be removing the links from this._senders and this._receivers here?  

---
In reply to: [139532134](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139532134) [](ancestors = 139532134)

Reply:
we want to forceDetach, not detach. if the connection has an error and has been disconnected amqp10 will crash if we call detach.

---
In reply to: [139532385](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139532385) [](ancestors = 139532385,139532134)

Reply:
we delete everything in _getRemainingLinks (put everything in one array and delete the references in the _senders and _receivers objects)

---
In reply to: [139532385](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139532385) [](ancestors = 139532385,139532134)